File: php/src/Output.php
<?php

declare(strict_types=1);

namespace FormHandle;

class Output
{
    private static bool $noColor = false;
    private static bool $initialized = false;

    private static function init(): void
    {
        if (!self::$initialized) {
            self::$noColor = getenv('NO_COLOR') !== false;
            self::$initialized = true;
        }
    }

    private static function c(string $code): string
    {
        self::init();
        return self::$noColor ? '' : $code;
    }

    public static function reset(): string { return self::c("\033[0m"); }
    public static function boldCode(): string { return self::c("\033[1m"); }
    public static function red(): string { return self::c("\033[31m"); }
    public static function green(): string { return self::c("\033[32m"); }
    public static function yellow(): string { return self::c("\033[33m"); }
    public static function blue(): string { return self::c("\033[34m"); }
    public static function cyan(): string { return self::c("\033[36m"); }
    public static function gray(): string { return self::c("\033[90m"); }

    public static function success(string $msg): void
    {
        echo self::green() . "\u{2714}" . self::reset() . " $msg\n";
    }

    public static function error(string $msg): void
    {
        fwrite(STDERR, self::red() . "\u{2716}" . self::reset() . " $msg\n");
    }

    public static function info(string $msg): void
    {
        echo self::blue() . "\u{2139}" . self::reset() . " $msg\n";
    }

    public static function warn(string $msg): void
    {
        echo self::yellow() . "\u{26a0}" . self::reset() . " $msg\n";
    }

    public static function dim(string $msg): string
    {
        return self::gray() . $msg . self::reset();
    }

    public static function bold(string $msg): string
    {
        return self::boldCode() . $msg . self::reset();
    }

    public static function heading(string $msg): void
    {
        echo "\n" . self::boldCode() . self::cyan() . $msg . self::reset() . "\n\n";
    }

    public static function json(mixed $data): void
    {
        echo json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES) . "\n";
    }

    public static function table(array $rows): void
    {
        if (empty($rows)) return;
        $maxKey = max(array_map(fn($r) => strlen($r[0]), $rows));
        foreach ($rows as [$key, $val]) {
            echo "  " . self::bold(str_pad($key, $maxKey)) . "  $val\n";
        }
    }
}

File: rust/src/output.rs
use std::env;

fn no_color() -> bool {
    env::var("NO_COLOR").is_ok()
}

fn c(code: &str) -> &str {
    if no_color() { "" } else { code }
}

pub fn success(msg: &str) {
    println!("{}\u{2714}{} {}", c("\x1b[32m"), c("\x1b[0m"), msg);
}

pub fn error(msg: &str) {
    eprintln!("{}\u{2716}{} {}", c("\x1b[31m"), c("\x1b[0m"), msg);
}

pub fn info(msg: &str) {
    println!("{}\u{2139}{} {}", c("\x1b[34m"), c("\x1b[0m"), msg);
}

pub fn warn(msg: &str) {
    println!("{}\u{26a0}{} {}", c("\x1b[33m"), c("\x1b[0m"), msg);
}

pub fn dim(msg: &str) -> String {
    format!("{}{}{}", c("\x1b[90m"), msg, c("\x1b[0m"))
}

pub fn bold(msg: &str) -> String {
    format!("{}{}{}", c("\x1b[1m"), msg, c("\x1b[0m"))
}

pub fn heading(msg: &str) {
    println!("\n{}{}{}{}\n", c("\x1b[1m"), c("\x1b[36m"), msg, c("\x1b[0m"));
}

pub fn json_out(value: &serde_json::Value) {
    println!("{}", serde_json::to_string_pretty(value).unwrap());
}

pub fn table(rows: &[(&str, &str)]) {
    if rows.is_empty() {
        return;
    }
    let max_key = rows.iter().map(|(k, _)| k.len()).max().unwrap_or(0);
    for (key, val) in rows {
        println!("  {}  {}", bold(&format!("{:<width$}", key, width = max_key)), val);
    }
}
